back-end/apis/model.py

```diff
@@ -139,160 +138,130 @@ def UploadModel():
-                code_path, class_name, RServer.get_model_wrapper().device
-            )
-        except Exception as e:
-            traceback.print_exc()
-            clear_model_temp_files(base_dir, saving_id)
-            return RResponse.abort(400, f"Failed to initialize the custom model. {e}")
-    elif "pretrained" in metadata:  # If the model is predefined
-        pretrained = bool(int(metadata.get("pretrained")))
-        num_classes = int(metadata.get("num_classes"))
-        try:
-            model = init_predefined_model(class_name, pretrained, num_classes)
-            with open(code_path, "w") as code_file:
-                code_file.write(f"num_classes = {num_classes}")
-        except Exception as e:
-            traceback.print_exc()
-            return RResponse.abort(
-                400, f"Failed to initialize the predefined model. {e}"
-            )
-    else:
-        return RResponse.abort(400, "The model is neither custom nor predefined.")
+        code_path = os.path.join(
+            RServer.get_server().base_dir,
+            "generated",
+            "models",
+            "code",
+            f"{saving_id}.py",
+        )
+        weight_path = os.path.join(
+            RServer.get_server().base_dir,
+            "generated",
+            "models",
+            "ckpt",
+            f"{saving_id}.pth",
+        )
 
-    # Get the weight file and save it to a temporary location if it exists
-    if "weight_file" in request.files:
-        weight_file = request.files.get("weight_file")
-        try:
-            weight_path = os.path.join(
-                RServer.get_server().base_dir, "generated", "models", f"{saving_id}.pth"
+        # Get the model's class name
+        class_name = metadata.get("class_name")
+
+        predefined = bool(int(metadata.get("predefined")))
+
+        # Save the model's code definition and initialize the model
+        if not predefined:  # If the model is custom
+            # Get the model definition code and save it to a temporary file
+            code = request.form.get("code")
+            save_code(code, code_path)
+            # Initialize the model
+            try:
+                model = RModelWrapper.init_custom_model(
+                    code_path, class_name, RServer.get_model_wrapper().device
+                )
+            except Exception as e:
+                traceback.print_exc()
+                clear_model_temp_files(code_path, weight_path)
+                return RResponse.fail(
+                    f"Failed to initialize the custom model. {e}", 400
+                )
+        elif predefined:  # If the model is predefined
+            pretrained = bool(int(metadata.get("pretrained")))
+            num_classes = int(metadata.get("num_classes"))
+            # TODO: Stop relying on the user to provide the number of classes
+            code = f"num_classes = {num_classes}"
+            save_code(code, code_path)
```

Comment:
I think for pre-defined models we don't need to save any code right? Looks like you are trying to save some dummy string which is never used? Shall we remove?

Reply:
Yes, I think after initializing this `num_classes` in RDataManager, we can remove this. But currently `load_model_by_name` still relies on this (I changed the code in `load_model_by_name` to read this variable from the code file instead of using `self.num_classes` for a correct loading of the predefined model to pass my manual test).

Reply:
I will set this `num_classes` in `RDataManager` after this merge. I have created issue #207 for this.

Reply:
Got it. Thanks @Leon-Leyang